### PR TITLE
Fix remote arg processing

### LINKED ISSUE #2756: Significant latency upon remote evaluation
**Describe the bug**
When evaluating an expression with `nyxt --remote --eval`,
there is a significant delay on evaluation compared with evaluating through slime, for example.

**Precise recipe to reproduce the issue**
- Start nyxt
- Open multiple buffers
- Ensure `remote-execution-p` is set to `t`
- Run `nyxt --remote --eval '(switch-buffer-next)'` in a shell

**Information**
- OS name+version: Guix 5e4ec8218142eee8e6e148e787381a5ef891c5b1
- Graphics card and driver: Don't know
- Desktop environment / Window manager name+version: StumpWM 22.05
- How you installed Nyxt (Guix pack, package manager, build from source): `guix package -i nyxt`
- Information from `show-system-information`:
```
Nyxt version: 2.2.4
Renderer version: GI-GTK
Operating system kernel: Linux 5.19.16-gnu
Lisp implementation: SBCL 2.2.11 (Dynamic space size: 3221225472)
Features: (:WEBKIT2 :WEBKIT2-2.36 :WEBKIT2-CORS-ALLOWLIST :WEBKIT2-PASTE-PLAINTEXT
 :WEBKIT2-TRACKING :WEBKIT2-MUTE :WEBKIT2-EMOJI :WEBKIT2-MEDIA
 :WEBKIT2-SANDBOXING :GTK-3-22 :GTK-3-20 :GTK-3-18 :GTK-3-16 :GTK-3-14
 :GTK-3-12 :GTK-3-10 :GTK-3-8 :GTK-3-6 :GTK-3-4 :GTK :GDK-3-22 :GDK-3-20
 :GDK-3-18 :GDK-3-16 :GDK-3-14 :GDK-3-12 :GDK-3-10 :GDK-3-8 :GDK-3-6 :GDK-3-4
 :CAIRO-1-10 :CAIRO-1-12 :GDK-PIXBUF :GLIB-2-30 :GLIB-2-32 :GLIB-2-34
 :GLIB-2-36 :GLIB-2-38 :GLIB-2-40 :GLIB-2-42 :GLIB-2-44 :GLIB-2-46 :GLIB-2-48
 :GLIB-2-50 :GLIB-2-52 :GLIB-2-54 :GLIB-2-56 :GLIB-2-58 :GLIB :NYXT-2
 :FSET-EXT-STRINGS :CUSTOM-HASH-TABLE-NATIVE :CLIPBOARD-CONTENT-METHOD :SWANK
 :PLUMP-UTF-32 :GLOBAL-VARS :PARENSCRIPT :NAMED-READTABLES :LPARALLEL
 :21BIT-CHARS :CHUNGA :FLEXI-STREAMS :CLOSER-MOP :CL-PPCRE-UNICODE :CL-UNICODE
 :CL-PPCRE :BORDEAUX-THREADS :LPARALLEL.WITH-CLTL2 :LPARALLEL.WITH-CAS
 :LPARALLEL.WITH-STEALING-SCHEDULER :SPLIT-SEQUENCE CHIPZ-SYSTEM:GRAY-STREAMS
 CFFI-FEATURES:FLAT-NAMESPACE CFFI-FEATURES:X86-64 CFFI-FEATURES:UNIX :CFFI
 CFFI-SYS::FLAT-NAMESPACE ALEXANDRIA::SEQUENCE-EMPTYP :FAST-IO-SV :FAST-IO
 :SBCL-USES-SB-ROTATE-BYTE :ASDF-SYSTEM-CONNECTIONS :CL-JSON-CLOS :CL-JSON
 :THREAD-SUPPORT :ASDF3.3 :ASDF3.2 :ASDF3.1 :ASDF3 :ASDF2 :ASDF :OS-UNIX
 :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :ARENA-ALLOCATOR :X86-64 :GENCGC :64-BIT
 :ANSI-CL :COMMON-LISP :ELF :IEEE-FLOATING-POINT :LINUX :LITTLE-ENDIAN
 :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB :SB-PACKAGE-LOCKS
 :SB-THREAD :SB-UNICODE :SBCL :UNIX)

ASDF version: 3.3.5
ASDF registries: (NYXT-SOURCE-REGISTRY ENVIRONMENT-SOURCE-REGISTRY USER-SOURCE-REGISTRY
                      USER-SOURCE-REGISTRY-DIRECTORY
                      DEFAULT-USER-SOURCE-REGISTRY SYSTEM-SOURCE-REGISTRY
                      SYSTEM-SOURCE-REGISTRY-DIRECTORY
                      DEFAULT-SYSTEM-SOURCE-REGISTRY)
Critical dependencies: (/gnu/store/xrr6jnmk88mpadwmx9mlkdydl3s0bwh2-sbcl-cl-cffi-gtk-0.11.2-2.e9a46df/share/common-lisp/sbcl/cl-cffi-gtk/gtk/cl-cffi-gtk.asd
 /gnu/store/w64c12i7y7vcc9vdys0wj4vyqq7vfch7-sbcl-cl-gobject-introspection-0.3-1.d0136c8/share/common-lisp/sbcl/cl-gobject-introspection/cl-gobject-introspection.asd
 /gnu/store/91y87sr1zr2qdaxk6cfzsib6792jbahj-sbcl-cl-webkit-3.5.6/share/common-lisp/sbcl/cl-webkit/webkit2/cl-webkit2.asd)

Guix version:
```

More discussion here:
https://discourse.atlas.engineer/t/evaluating-expressions-from-terminal-as-if-evaluated-from-the-window/621/5

### PR DESCRIPTION
**Pull request by Ambrevar**
# Description

- Skip config files with `--remote`.
- Log socket path.
- Enable socket with `--failsafe --remote`.
- Allow multiple `--load` and `--eval` calls.
- Do not focus window with `--remote`.
- `--remote` now reads from standard input.

Fixes #2756, #2757.

# Discussion

@ErikPrantare In my test remote evaluations are very fast.  There is just the startup delay of the Nyxt executable (<500ms), that's it.

Can you report if you suffer any longer delay?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] I have pulled from master before submitting this PR
- [x] There are no merge conflicts.
- [x] I've added the new dependencies as:
  - [ ] ASDF dependencies,
  - [ ] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [ ] and Guix dependencies.
- [x] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [x] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [x] Documentation:
  - [x] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [x] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [ ] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [ ] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - [x] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [ ] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by ErikPrantare** [comment]
> @ErikPrantare In my test remote evaluations are very fast. There is just the startup delay of the Nyxt executable (<500ms), that's it.

My delay is not much greater than that,  but it is enough to be annoying for my use case, which is to control nyxt through the shell using voice commands. Then it becomes 500ms delay for every interactive action, which is significant. Copying from the discussion on the forum:

>What I am trying to do is getting voice control software (talon voice) to work smoothly with the browser, so I am having it execute commands through the shell. The current way I am doing it is having emacs connected to nyxt through the swank server and using slime through emacsclient --eval, but this necessitates both emacs running and it having an established connection to nyxt already. Letting talon communicate directly with nyxt and having it run as smoothly as evaluating the expressions through slime would be very nice.

**Comment by Ambrevar** [comment]
Can you post the shell output of

```sh
$ time nyxt --failsafe --eval '17' --quit
```

**Comment by ErikPrantare** [comment]
I have no `--failsafe`, but

```
$ time nyxt --no-init --no-auto-config --eval '17' --quit
17
nyxt --no-init --no-auto-config --eval '17' --quit  0.28s user 0.05s system 100% cpu 0.329 total
```

**Comment by Ambrevar** [comment]
Could try this branch see how it improves the business?
Also please run the `time ...` command twice and report the second result.

**Comment by Ambrevar** [comment]
@aartaka @aadcg Any suggestion how we can improve `--remote` latency?

**Comment by ErikPrantare** [comment]
> Also please run the `time ...` command twice and report the second result.

```
$ for i in `seq 10`; do; time nyxt --no-init --no-auto-config --remote --eval '17' --quit; done;
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.25s user 0.07s system 100% cpu 0.315 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.29s user 0.08s system 99% cpu 0.368 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.25s user 0.06s system 100% cpu 0.312 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.29s user 0.04s system 100% cpu 0.328 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.26s user 0.04s system 100% cpu 0.296 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.24s user 0.06s system 100% cpu 0.300 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.26s user 0.04s system 100% cpu 0.300 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.24s user 0.06s system 100% cpu 0.299 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.24s user 0.06s system 100% cpu 0.303 total
nyxt --no-init --no-auto-config --remote --eval '17' --quit  0.25s user 0.05s system 100% cpu 0.305 total
```

**Comment by aartaka** [comment]
> @aartaka @aadcg Any suggestion how we can improve `--remote` latency?

If that's the executable startup latency, the best one can do is recompile with less SBCL compression, it seems...

**Comment by Ambrevar** [comment]
Compression is off by default.

**Comment by Ambrevar** [comment]
@ErikPrantare Can you try with this PR branch?

**Comment by aadcg** [comment]
@Ambrevar here are my local results after building nyxt from this branch:

```log
$ time ./nyxt --failsafe --eval '17' --quit
17

real	0m0.388s
user	0m0.283s
sys	0m0.108s
$ time ./nyxt --failsafe --eval '17' --quit
17

real	0m0.393s
user	0m0.294s
sys	0m0.101s
```

**Comment by Ambrevar** [comment]
Thanks for sharing, I have similar results.
It may be cl-cffi-gtk that's slowing down the startup so much.

**Comment by ErikPrantare** [comment]
> @ErikPrantare Can you try with this PR branch?

I am afraid I have too little know-how in this area to know how to do that.

---

How does emacsclient manage to evaluate stuff so quickly? Perhaps inspiration can be found in the implementation of `emacsclient --eval`

**Comment by Ambrevar** [comment]
> How does emacsclient manage to evaluate stuff so quickly? Perhaps inspiration can be found in the implementation of `emacsclient --eval`

That's one solution indeed: write our own `nyxtclient`.  Which would be a trivial program that would solely execute `remote-eval`.  It's just a bit of work in terms of deployment, and I wonder if there is a simpler solution...

**Comment by Ambrevar** [comment]
I've implemented `nyxt-client`.  You can build it with `make nyxt-client`.

It runs faster by an order of magnitude.  That said, it may not be optimally fast still.
Plus it takes lots of disk space.  Any better idea?

Maybe we could use a different compiler for `nyxt-client`, like CLISP.

File: nyxt.asd [hunk 1]
```diff
@@ -513,6 +514,28 @@ The renderer is configured from NYXT_RENDERER or `*nyxt-renderer*'."))
                         (call-next-method)
                         (warn "`nyxt' executable missing, skipping tests."))))
 
+(defsystem "nyxt-client"
```
Comment (aartaka) [inline nyxt.asd:517]: Should probably be called `nyxt/client`, otherwise ASDF is throwing warnings.

**Review by aartaka — APPROVED** [review]

**Comment by Ambrevar** [comment]
Maybe a better idea: Forget about nyxt-client and instead have `nyxt --remote` _not quit_ and read from stdin (until Ctrl-D is pressed).

This mimicks the behaviour of most interpreters (`sbcl`, `guix repl -t machine`, etc.).

**Comment by Ambrevar** [comment]
Done!  bd1e8a14848d7f0d847059345e27c82894e4c307

This is awesome, we now have a Nyxt shell interpreter!

I believe this PR is ready to merge.

File: nyxt.asd (hunk 1 above)
Comment (Ambrevar) [reply nyxt.asd]: Got rid of it anyways.

**Comment by ErikPrantare** [comment]
Nice!

**Pull request merged**